src/OrleansEventSourcing/Common/PrimaryBasedLogViewAdaptor.cs

```diff
@@ -0,0 +1,895 @@
+        }
+        /// <summary>
+        /// Wait until this cluster has received a configuration that is at least as new as timestamp
+        /// </summary>
+        protected async Task GetCaughtUpWithConfigurationAsync(DateTime adminTimestamp)
+        {
+            while (Configuration == null || Configuration.AdminTimestamp < adminTimestamp)
+            {
+                Services.Verbose("Waiting for config {0}", adminTimestamp);
+
+                await Task.Delay(5000);
+            }
+        }
+
+
+
+
+
+        #region Interface
+
+        /// <inheritdoc />
+        public void Submit(TLogEntry logEntry)
+        {
+            if (!SupportSubmissions)
+                throw new InvalidOperationException("provider does not support submissions on cluster " + Services.MyClusterId);
+
+            if (stats != null) stats.EventCounters["SubmitCalled"]++;
+
+            Services.Verbose2("Submit");
+
+            SubmitInternal(DateTime.UtcNow, logEntry);
+
+            worker.Notify();
+        }
+
+        /// <inheritdoc />
+        public void SubmitRange(IEnumerable<TLogEntry> logEntries)
+        {
+            if (!SupportSubmissions)
+                throw new InvalidOperationException("Provider does not support submissions on cluster " + Services.MyClusterId);
+
+            if (stats != null) stats.EventCounters["SubmitRangeCalled"]++;
+
+            Services.Verbose2("SubmitRange");
+
+            var time = DateTime.UtcNow;
+
+            foreach (var e in logEntries)
+                SubmitInternal(time, e);
+
+            worker.Notify();
+        }
+
+        /// <inheritdoc />
+        public Task<bool> TryAppend(TLogEntry logEntry)
+        {
+            if (!SupportSubmissions)
+                throw new InvalidOperationException("Provider does not support submissions on cluster " + Services.MyClusterId);
+
+            if (stats != null) stats.EventCounters["TryAppendCalled"]++;
+
+            Services.Verbose2("TryAppend");
+
+            var promise = new TaskCompletionSource<bool>();
+
+            SubmitInternal(DateTime.UtcNow, logEntry, GetConfirmedVersion() + pending.Count, promise);
+
+            worker.Notify();
+
+            return promise.Task;
+        }
+
+        /// <inheritdoc />
+        public Task<bool> TryAppendRange(IEnumerable<TLogEntry> logEntries)
+        {
+            if (!SupportSubmissions)
+                throw new InvalidOperationException("Provider does not support submissions on cluster " + Services.MyClusterId);
+
+            if (stats != null) stats.EventCounters["TryAppendRangeCalled"]++;
+
+            Services.Verbose2("TryAppendRange");
+
+            var promise = new TaskCompletionSource<bool>();
+            var time = DateTime.UtcNow;
+            var pos = GetConfirmedVersion() + pending.Count;
+
+            bool first = true;
+            foreach (var e in logEntries)
+            {
+                SubmitInternal(time, e, pos++, first ? promise : null);
```

Comment:
Should the promise be specified on the first or the last entry? I would have assumed the last, so that the caller is notified on successful submission of the entire range of entries.